File: src/lib/sandbox/agenticfactor/gmail.py
"""
AgenticFactor SDK — Gmail Module
Send, read, search, draft emails via Gmail API using stored OAuth tokens.
"""

import base64
import json
import os
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.mime.base import MIMEBase
from email import encoders
from typing import Optional, List, Dict, Any

from ._core import _get_token, _request

GMAIL_API = "https://gmail.googleapis.com/gmail/v1/users/me"


def _token():
    return _get_token("google")


def send(
    to: str,
    subject: str,
    body: str,
    cc: Optional[str] = None,
    bcc: Optional[str] = None,
    html: bool = False,
    reply_to: Optional[str] = None,
    attachments: Optional[List[Dict[str, Any]]] = None,
) -> Dict:
    """
    Send an email via Gmail API.
    
    Args:
        to: Recipient email (comma-separated for multiple)
        subject: Email subject line
        body: Email body (plain text or HTML)
        cc: CC recipients (comma-separated)
        bcc: BCC recipients (comma-separated)
        html: If True, send as HTML email
        reply_to: Message-ID to reply to
        attachments: List of {"filename": "...", "content": bytes, "mime_type": "..."}
    
    Returns:
        Dict with message id and thread id
    """
    token = _token()
    
    if attachments:
        msg = MIMEMultipart()
        if html:
            msg.attach(MIMEText(body, "html"))
        else:
            msg.attach(MIMEText(body, "plain"))
        
        for att in attachments:
            part = MIMEBase("application", "octet-stream")
            content = att.get("content", b"")
            if isinstance(content, str):
                content = content.encode()
            part.set_payload(content)
            encoders.encode_base64(part)
            part.add_header("Content-Disposition", f'attachment; filename="{att.get("filename", "file")}"')
            msg.attach(part)
    else:
        content_type = "html" if html else "plain"
        msg = MIMEText(body, content_type)
    
    msg["to"] = to
    msg["subject"] = subject
    if cc:
        msg["cc"] = cc
    if bcc:
        msg["bcc"] = bcc
    if reply_to:
        msg["In-Reply-To"] = reply_to
        msg["References"] = reply_to
    
    raw = base64.urlsafe_b64encode(msg.as_bytes()).decode("utf-8")
    
    result = _request(
        "POST",
        f"{GMAIL_API}/messages/send",
        token=token,
        json_data={"raw": raw},
        provider="gmail",
    )
    return {"id": result.get("id"), "threadId": result.get("threadId"), "status": "sent"}


def read(message_id: str) -> Dict:
    """Read a specific email by message ID."""
    token = _token()
    result = _request(
        "GET",
        f"{GMAIL_API}/messages/{message_id}",
        token=token,
        params={"format": "full"},
        provider="gmail",
    )
    
    headers = {h["name"].lower(): h["value"] for h in result.get("payload", {}).get("headers", [])}
    
    # Extract body
    body = ""
    payload = result.get("payload", {})
    if payload.get("body", {}).get("data"):
        body = base64.urlsafe_b64decode(payload["body"]["data"]).decode("utf-8", errors="replace")
    elif payload.get("parts"):
        for part in payload["parts"]:
            if part.get("mimeType") in ["text/plain", "text/html"]:
                data = part.get("body", {}).get("data", "")
                if data:
                    body = base64.urlsafe_b64decode(data).decode("utf-8", errors="replace")
                    break
    
    # Extract attachments info
    attachments = []
    for part in payload.get("parts", []):
        if part.get("filename"):
            attachments.append({
                "filename": part["filename"],
                "mimeType": part.get("mimeType"),
                "size": part.get("body", {}).get("size", 0),
                "attachmentId": part.get("body", {}).get("attachmentId"),
            })
    
    return {
        "id": result.get("id"),
        "threadId": result.get("threadId"),
        "from": headers.get("from", ""),
        "to": headers.get("to", ""),
        "subject": headers.get("subject", ""),
        "date": headers.get("date", ""),
        "body": body,
        "snippet": result.get("snippet", ""),
        "labels": result.get("labelIds", []),
        "attachments": attachments,
    }


def search(
    query: str,
    max_results: int = 10,
    label: Optional[str] = None,
) -> List[Dict]:
    """
    Search emails in Gmail.
    
    Args:
        query: Gmail search query (e.g., "from:user@example.com subject:invoice has:attachment")
        max_results: Maximum number of results (default 10)
        label: Filter by label (e.g., "INBOX", "SENT", "DRAFT")
    
    Returns:
        List of email summaries
    """
    token = _token()
    params = {"q": query, "maxResults": min(max_results, 50)}
    if label:
        params["labelIds"] = label
    
    result = _request(
        "GET",
        f"{GMAIL_API}/messages",
        token=token,
        params=params,
        provider="gmail",
    )
    
    messages = result.get("messages", [])
    emails = []
    for msg in messages[:max_results]:
        try:
            email_data = read(msg["id"])
            emails.append(email_data)
        except Exception as e:
            emails.append({"id": msg["id"], "error": str(e)})
    
    return emails


def draft(
    to: str,
    subject: str,
    body: str,
    html: bool = False,
) -> Dict:
    """Create a draft email (not sent)."""
    token = _token()
    
    content_type = "html" if html else "plain"
    msg = MIMEText(body, content_type)
    msg["to"] = to
    msg["subject"] = subject
    
    raw = base64.urlsafe_b64encode(msg.as_bytes()).decode("utf-8")
    
    result = _request(
        "POST",
        f"{GMAIL_API}/drafts",
        token=token,
        json_data={"message": {"raw": raw}},
        provider="gmail",
    )
    return {"id": result.get("id"), "status": "draft_created"}


def list_labels() -> List[Dict]:
    """List all Gmail labels."""
    token = _token()
    result = _request("GET", f"{GMAIL_API}/labels", token=token, provider="gmail")
    return result.get("labels", [])


def download_attachment(message_id: str, attachment_id: str) -> bytes:
    """Download an email attachment."""
    token = _token()
    result = _request(
        "GET",
        f"{GMAIL_API}/messages/{message_id}/attachments/{attachment_id}",
        token=token,
        provider="gmail",
    )
    data = result.get("data", "")
    return base64.urlsafe_b64decode(data)

File: src/lib/services/tools/gmail.ts
import { ToolExecutionContext, registerTool } from './index';
import { createServiceClient } from '@/lib/supabase/server';

const GMAIL = 'https://gmail.googleapis.com/gmail/v1/users/me';

async function getToken(tenantId: string): Promise<string | null> {
  const supabase = createServiceClient();
  const { data } = await supabase
    .from('tenant_permissions')
    .select('access_token')
    .eq('tenant_id', tenantId)
    .eq('provider', 'google')
    .single();
  return data?.access_token ?? null;
}

function noCredError() {
  return { error: 'Google not connected. Please connect Google in the Connectors page.', connector_required: true, provider: 'google', connection_type: 'oauth' };
}

function makeRfc2822(to: string | string[], subject: string, body: string, from?: string, cc?: string, replyTo?: string): string {
  const toLine = Array.isArray(to) ? to.join(', ') : to;
  const headers = [`To: ${toLine}`, `Subject: ${subject}`, 'MIME-Version: 1.0', 'Content-Type: text/html; charset=UTF-8'];
  if (from) headers.unshift(`From: ${from}`);
  if (cc) headers.push(`Cc: ${cc}`);
  if (replyTo) headers.push(`Reply-To: ${replyTo}`);
  return Buffer.from(headers.join('\r\n') + '\r\n\r\n' + body).toString('base64url');
}

async function sendEmailTool({ tenantId, args }: ToolExecutionContext) {
  const token = await getToken(tenantId);
  if (!token) return noCredError();
  const { to, subject, body, from, cc, reply_to, thread_id } = args as {
    to: string | string[]; subject: string; body: string;
    from?: string; cc?: string; reply_to?: string; thread_id?: string;
  };
  if (!to || !subject || !body) return { error: 'to, subject and body are required' };
  const raw = makeRfc2822(to, subject, body, from, cc, reply_to);
  const payload: Record<string, string> = { raw };
  if (thread_id) payload.threadId = thread_id;
  const res = await fetch(`${GMAIL}/messages/send`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!res.ok) return { error: `Gmail error: HTTP ${res.status}` };
  const data = await res.json() as { id: string; threadId: string };
  return { message_id: data.id, thread_id: data.threadId };
}

async function readInboxTool({ tenantId, args }: ToolExecutionContext) {
  const token = await getToken(tenantId);
  if (!token) return noCredError();
  const { max_results = 10, query = 'in:inbox', label_ids } = args as { max_results?: number; query?: string; label_ids?: string[] };
  const params = new URLSearchParams({ maxResults: String(Math.min(max_results, 50)), q: query });
  if (label_ids?.length) params.set('labelIds', label_ids.join(','));
  const listRes = await fetch(`${GMAIL}/messages?${params}`, { headers: { Authorization: `Bearer ${token}` } });
  if (!listRes.ok) return { error: `Gmail error: HTTP ${listRes.status}` };
  const listData = await listRes.json() as { messages?: Array<{ id: string }> };
  if (!listData.messages?.length) return { emails: [], total: 0 };

  const emails = await Promise.all(listData.messages.slice(0, 10).map(async m => {
    const msgRes = await fetch(`${GMAIL}/messages/${m.id}?format=metadata&metadataHeaders=From&metadataHeaders=To&metadataHeaders=Subject&metadataHeaders=Date`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (!msgRes.ok) return null;
    const msg = await msgRes.json() as { id: string; threadId: string; snippet: string; payload: { headers: Array<{ name: string; value: string }> } };
    const h = Object.fromEntries(msg.payload.headers.map(hdr => [hdr.name.toLowerCase(), hdr.value]));
    return { id: msg.id, thread_id: msg.threadId, from: h.from, to: h.to, subject: h.subject, date: h.date, snippet: msg.snippet };
  }));

  return { emails: emails.filter(Boolean), total: listData.messages.length };
}

async function searchEmailsTool({ tenantId, args }: ToolExecutionContext) {
  const token = await getToken(tenantId);
  if (!token) return noCredError();
  const { query, max_results = 20 } = args as { query: string; max_results?: number };
  if (!query) return { error: 'query is required (Gmail search syntax)' };
  const params = new URLSearchParams({ maxResults: String(Math.min(max_results, 100)), q: query });
  const res = await fetch(`${GMAIL}/messages?${params}`, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) return { error: `Gmail error: HTTP ${res.status}` };
  const data = await res.json() as { messages?: Array<{ id: string }>; resultSizeEstimate?: number };
  return { message_ids: data.messages?.map(m => m.id) ?? [], estimated_total: data.resultSizeEstimate ?? 0 };
}

async function createDraftTool({ tenantId, args }: ToolExecutionContext) {
  const token = await getToken(tenantId);
  if (!token) return noCredError();
  const { to, subject, body, from } = args as { to: string; subject: string; body: string; from?: string };
  if (!to || !subject || !body) return { error: 'to, subject and body are required' };
  const raw = makeRfc2822(to, subject, body, from);
  const res = await fetch(`${GMAIL}/drafts`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ message: { raw } }),
  });
  if (!res.ok) return { error: `Gmail error: HTTP ${res.status}` };
  const data = await res.json() as { id: string };
  return { draft_id: data.id, to, subject };
}

registerTool('gmail_send', sendEmailTool);
registerTool('gmail_read_inbox', readInboxTool);
registerTool('gmail_search', searchEmailsTool);
registerTool('gmail_create_draft', createDraftTool);
